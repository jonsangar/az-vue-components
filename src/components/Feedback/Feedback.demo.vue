<template lang="pug">
  .feedbackDemo
    Demo(title="Uso Básico")
      az-feedback(:isLoading="isLoading" :isSuccess="isSuccess" :isError="isError")

      .text-center
        az-button.margin-right-half(@click="onLoading" size="small") Loading
        az-button.margin-right-half(@click="onSuccess" type="success" size="small") Success
        az-button(@click="onError" type="danger" size="small") Error

    Demo(title="Uso Básico")
      az-feedback(:isLoading="isLoading2" :isSuccess="isSuccess2" :isError="isError2" :scale="1")

      .text-center
        az-button.margin-right-half(@click="onLoading2" size="small") Loading
        az-button.margin-right-half(@click="onSuccess2" type="success" size="small") Success
        az-button(@click="onError2" type="danger" size="small") Error
</template>

<script>
export default {
  data() {
    return {
      isLoading: true,
      isSuccess: false,
      isError: false,
      isLoading2: true,
      isSuccess2: false,
      isError2: false,
    };
  },
  methods: {
    onLoading() {
      this.isLoading = true;
      this.isSuccess = false;
      this.isError = false;
    },

    onSuccess() {
      this.isLoading = false;
      this.isSuccess = true;
      this.isError = false;
    },

    onError() {
      this.isLoading = false;
      this.isSuccess = false;
      this.isError = true;
    },

    onLoading2() {
      this.isLoading2 = true;
      this.isSuccess2 = false;
      this.isError2 = false;
    },

    onSuccess2() {
      this.isLoading2 = false;
      this.isSuccess2 = true;
      this.isError2 = false;
    },

    onError2() {
      this.isLoading2 = false;
      this.isSuccess2 = false;
      this.isError2 = true;
    },
  },
};
</script>

<style lang="scss">
.az-feedback {
  margin: double() auto;
}
</style>
