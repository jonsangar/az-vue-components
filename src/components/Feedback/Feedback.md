---
device: true
---

# Feedback Loader

### Instalación

```javascript
import { Feedback } from 'didor';

Vue.use(Feedback);
```

---

### Uso

#### Uso básico

``` html
<az-feedback :isLoading="isLoading" :isSuccess="isSuccess" :isError="isError">
```

#### Uso escalas

``` html
<az-feedback :scale=1>
```

---

### API

| Attribute   | Description                                | Type      | Default   |
| ----------- | ------------------------------------------ | --------- | --------- |
| isLoading   | Cambia al estado loading                   | `Boolean` | `false`   |
| isSuccess   | Cambia al estado correcto                  | `Boolean` | `false`   |
| isError     | Cambia al estado error                     | `Boolean` | `false`   |
| scale       | Cambia la escala                           | `Integer` | 0.5       |
