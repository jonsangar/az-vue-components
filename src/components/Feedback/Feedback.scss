$feedbackBackground: #ffffff !default;
$feedbackSuccess: color(success) !default;
$feedbackSuccessLight: color(success-light) !default;
$feedbackError: color(danger) !default;
$feedbackErrorLight: color(danger-light) !default;
$feedbackBorderColor: color(gray6) !default;
$feedbackSpinnerColor: color(brand) !default;
$feedbackSpinnerWidth: 2px !default;
$feedbackMarkColor: #ffffff !default;

.az-feedback {
  $rootClass: &;

  position: relative;
  width: 107px;
  height: 107px;
  user-select: none;
  background-color: $feedbackBackground;
  border: 8px solid $feedbackBorderColor;
  border-radius: 50%;

  //
  // Background circles initial state
  //
  &::after,
  &::before {
    position: absolute;
    top: -4px;
    left: -4px;
    display: none;
    width: 102px;
    height: 102px;
    content: '';
    border-radius: 50%;
    opacity: 0;
    transform: scale(0);
  }

  &__spinner {
    position: absolute;
    top: -9px;
    left: -9px;
    width: 110px;
    height: 110px;

    svg {
      display: none;
      width: 100%;
      height: 100%;
      animation: feedback-spinner 1.2s infinite linear;
    }

    circle {
      fill: none;
      stroke: $feedbackSpinnerColor;
      stroke-dasharray: 65, 250;
      stroke-width: 8;
      transition: all 800ms ease;
    }
  }

  //
  // Hide Marks
  //
  &__successMark,
  &__errorMark {
    display: none;
  }

  //
  // Success Mark
  //
  &__successMark {
    &::after {
      position: absolute;
      top: 50px;
      left: 19px;
      z-index: 1;
      width: 30px;
      height: 50px;
      content: '';
      border-top: 12px solid $feedbackMarkColor;
      border-right: 12px solid $feedbackMarkColor;
      transform: scaleX(-1) rotate(135deg);
      transform-origin: left top;
    }
  }

  //
  // Error Mark
  //
  &__errorMark {
    &::after,
    &::before {
      position: absolute;
      top: 43px;
      left: 22px;
      z-index: 1;
      width: 50px;
      height: 10px;
      content: '';
      background-color: $feedbackMarkColor;
      transform-origin: center center;
    }

    &::after {
      transform: rotate(135deg);
    }

    &::before {
      transform: rotate(45deg);
    }
  }

  //
  // Loading State
  //
  &--loading {
    svg {
      display: block;
    }
  }

  //
  // Success State
  //
  &--success {
    animation: feedback 1s forwards;
    animation-delay: 800ms;

    svg {
      display: block;

      circle {
        stroke: $feedbackSuccess;
        stroke-dasharray: 315, 250;
      }
    }

    &::after {
      display: block;
      background-color: $feedbackSuccess;
      animation: feedback-after 1s forwards;
      animation-delay: 800ms;
    }

    &::before {
      display: block;
      background-color: $feedbackSuccessLight;
      animation: feedback-before 1s forwards;
      animation-delay: 800ms;
    }

    #{$rootClass}__successMark {
      display: block;
    }
  }

  //
  // Error State
  //
  &--error {
    animation: shake 1s forwards;
    animation-delay: 800ms;

    svg {
      display: block;

      circle {
        stroke: $feedbackError;
        stroke-dasharray: 315, 250;
      }
    }

    &::after {
      display: block;
      background-color: $feedbackError;
      animation: feedback-after 1s forwards;
      animation-delay: 800ms;
    }

    &::before {
      display: block;
      background-color: $feedbackErrorLight;
      animation: feedback-before 1s forwards;
      animation-delay: 800ms;
    }

    #{$rootClass}__errorMark {
      display: block;
    }
  }
}

@keyframes feedback-spinner {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}

@keyframes feedback {
  0% {
    transform: scale(1);
  }
  15% {
    transform: scale(0.9);
  }
  30% {
    transform: scale(1.05);
  }
  60% {
    transform: scale(1);
  }
  100% {
    transform: scale(1);
  }
}

@keyframes shake {
  10%,
  90% {
    transform: translate3d(-1px, 0, 0);
  }

  20%,
  80% {
    transform: translate3d(2px, 0, 0);
  }

  30%,
  50%,
  70% {
    transform: translate3d(-4px, 0, 0);
  }

  40%,
  60% {
    transform: translate3d(4px, 0, 0);
  }
}

@keyframes feedback-after {
  0% {
    opacity: 0;
    transform: scale(0);
  }
  30% {
    opacity: 0;
    transform: scale(0);
  }
  60% {
    opacity: 1;
    transform: scale(1);
  }
  100% {
    opacity: 1;
    transform: scale(1);
  }
}

@keyframes feedback-before {
  0% {
    opacity: 0;
    transform: scale(0);
  }
  15% {
    opacity: 0;
    transform: scale(0);
  }
  40% {
    opacity: 1;
    transform: scale(1);
  }
  60% {
    opacity: 1;
    transform: scale(1);
  }
  61% {
    opacity: 0;
    transform: scale(1);
  }
  100% {
    opacity: 0;
    transform: scale(1);
  }
}
