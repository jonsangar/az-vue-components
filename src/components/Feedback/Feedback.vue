<template lang="pug">
  .az-feedback__scale(:style="{'transform': style}")
    .az-feedback(:class="classObject")
      .az-feedback__spinner
        svg
          circle(cx="50%" cy="50%" r="50px")
      .az-feedback__successMark
      .az-feedback__errorMark
</template>

<script>
export default {
  props: {
    isLoading: {
      type: Boolean,
      default: false,
    },
    isSuccess: {
      type: Boolean,
      default: false,
    },
    isError: {
      type: Boolean,
      default: false,
    },
    scale: {
      type: Number,
      default: 0.5,
    },
  },
  computed: {
    classObject() {
      return [
        this.isLoading ? 'az-feedback--loading' : '',
        this.isSuccess ? 'az-feedback--success' : '',
        this.isError ? 'az-feedback--error' : '',
      ];
    },

    style() {
      return `scale(${this.scale})`;
    },
  },
};
</script>

<style src="./Feedback.scss" lang="scss"></style>
