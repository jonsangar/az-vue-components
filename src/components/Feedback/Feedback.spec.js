import { shallowMount, createLocalVue } from '@vue/test-utils';
// Importamos librerÃ­as de estilos e iconos
import SvgIcon from 'vue-svgicon'
import Feedback from './Feedback.vue';

const localVue = createLocalVue()
localVue.use(SvgIcon, {
  tagName: 'az-icon',
  classPrefix: 'az',
})

describe('Feedback.vue', () => {
  const build = () => {
    const wrapper = shallowMount(Feedback, {
      localVue,
    })

    return {
      wrapper,
    }
  }

  it('Comprueba renderizado del componente Feedback con snapshot', () => {
    const { wrapper } = build()
    expect(wrapper.html()).toMatchSnapshot()

  })
})
