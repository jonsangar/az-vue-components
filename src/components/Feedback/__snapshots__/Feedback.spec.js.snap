// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feedback.vue Comprueba renderizado del componente Feedback con snapshot 1`] = `
<div class="az-feedback__scale" style="transform: scale(0.5);">
  <div class="az-feedback">
    <div class="az-feedback__spinner"><svg>
        <circle cx="50%" cy="50%" r="50px"></circle>
      </svg></div>
    <div class="az-feedback__successMark"></div>
    <div class="az-feedback__errorMark"></div>
  </div>
</div>
`;
